input/input.go

```diff
@@ -72,3 +92,99 @@ func (in DefaultHandler) Process(metric *schema.MetricData, partition int32) {
 	m.Add(uint32(metric.Time), metric.Value)
 	in.pressureTank.Add(int(time.Since(pre).Nanoseconds()))
 }
+
+func (in *DefaultHandler) SetInterval(metric *schema.MetricData, partition int32) {
+	b, ok := in.buffer.Load(metric.Id)
+	if !ok {
+		dp := make([]*schema.MetricData, 1, 3)
+		dp[0] = metric
+		in.buffer.Store(metric.Id, &IntervalLookupRecord{
+			DataPoints: dp,
+		})
+		return
+	}
+	ilr := b.(*IntervalLookupRecord)
+	ilr.Lock()
+	// check for duplicate TS
+	if ilr.DataPoints[ilr.pos].Time == metric.Time {
+		//drop the metric as it is a duplicate.
+		ilr.Unlock()
+		return
+	}
+
+	// add this datapoint to our circular buffer
+	if len(ilr.DataPoints) < 3 {
+		ilr.DataPoints = append(ilr.DataPoints, metric)
+		ilr.pos++
+	} else {
+		if ilr.pos == 2 {
+			ilr.pos = 0
+		} else {
+			ilr.pos++
+		}
+		ilr.DataPoints[ilr.pos] = metric
+	}
+
+	// if the interval is already known and was updated in the last 24hours, use it.
+	if ilr.Interval != 0 && time.Since(ilr.Last) > 84600 {
+		metric.Interval = ilr.Interval
+		metric.SetId()
+		in.Process(metric, partition)
+		ilr.Unlock()
+		return
+	}
+
+	if len(ilr.DataPoints) < 3 {
+		// we dont have 3 points yet.
+		ilr.Unlock()
+		return
+	}
+	log.Debug("input: calculating interval of %s", metric.Id)
+
+	delta1 := ilr.DataPoints[1].Time - ilr.DataPoints[0].Time
+	// make sure the points are not out of order
+	if delta1 < 0 {
+		delta1 = -1 * delta1
+	}
+	// make sure the points are not out of order
+	delta2 := ilr.DataPoints[2].Time - ilr.DataPoints[1].Time
+	if delta2 < 0 {
+		delta2 = -1 * delta2
+	}
+	interval := 0
+
+	// To protect against dropped metrics and out of order metrics, use the smallest delta as the interval.
+	// Because we have 3 datapoints, it doesnt matter what order they are in.  The smallest delta is always
+	// going to be correct. (unless their are out of order and dropped metrics)
```

Comment:
to be more precise, the out-of-orderness must be constrained to the window of 3, for the smallest delta to be correct.
consider this sequence of timestamps:
40, 60, 80, 30, 50, 70

any window of 3 will result in an outcome of 20s interval, which is not correct.
note that it's common to have out-of-orderness beyond 3 points. (default reorderbuffer is 20 which I think is not always enough)

note also that currently the reorderbuffer requires the interval to be known for it to properly bucketize values.

i want to think a bit more about how we can make the reorderbuffer work better together with interval detection.
especially because interval detection can work better if it's done after data is reordered.
(maybe a separate ROB implementation for unknown-interval data?)